plutarch-orphanage/src/PlutusLedgerApi/Orphans/Common.hs

```diff
@@ -228,3 +232,44 @@ instance (CoArbitrary k, CoArbitrary v) => CoArbitrary (AssocMap.Map k v) where
 instance (Function k, Function v) => Function (AssocMap.Map k v) where
   {-# INLINEABLE function #-}
   function = functionMap AssocMap.toList AssocMap.unsafeFromList
+
+-- | @since WIP
+newtype UnsortedAssocMap k v = UnsortedAssocMap (AssocMap.Map k v)
+  deriving newtype (Show, Eq, Ord, Pretty)
+
+-- | @since WIP
+instance (Arbitrary k, Ord k) => Arbitrary1 (UnsortedAssocMap k) where
+  {-# INLINEABLE liftArbitrary #-}
+  liftArbitrary genVal =
+    UnsortedAssocMap . AssocMap.unsafeFromList <$> do
+      keyList <- Set.toList <$> arbitrary
+      unsortedKeyList <- shuffle keyList
+      traverse (\key -> (key,) <$> genVal) unsortedKeyList
+
+  {-# INLINEABLE liftShrink #-}
+  liftShrink shrinkVal (UnsortedAssocMap aMap) =
+    UnsortedAssocMap . AssocMap.unsafeFromList <$> do
+      let asList = AssocMap.toList aMap
+      liftShrink (\(key, val) -> (key,) <$> shrinkVal val) asList
+
+-- | @since WIP
+instance (Arbitrary k, Arbitrary v, Ord k) => Arbitrary (UnsortedAssocMap k v) where
+  {-# INLINEABLE arbitrary #-}
+  arbitrary = liftArbitrary arbitrary
+
+  {-# INLINEABLE shrink #-}
+  shrink = liftShrink shrink
+
+-- | @since WIP
+instance (CoArbitrary k, CoArbitrary v) => CoArbitrary (UnsortedAssocMap k v) where
+  {-# INLINEABLE coarbitrary #-}
+  coarbitrary (UnsortedAssocMap aMap) = coarbitrary aMap
+
+-- | @since WIP
+instance (Function k, Function v) => Function (UnsortedAssocMap k v) where
+  {-# INLINEABLE function #-}
+  function = functionMap @(AssocMap.Map k v) coerce coerce
+
+-- | @since WIP
+getUnsortedAssocMap :: UnsortedAssocMap k v -> AssocMap.Map k v
+getUnsortedAssocMap = coerce
```

Comment:
We'll need to ensure our upstream contribution includes this.

Reply:
This was upstreamed in [IntersectMBO/plutus@`79d543c` (#6624)](https://github.com/IntersectMBO/plutus/pull/6624/commits/79d543ccfb797b2bd9e8b36e8ab400f54eed3d1a)